stdsimd-verify/tests/x86-intel.rs

```diff
@@ -0,0 +1,299 @@
+        (&Type::PrimUnsigned(64), "unsigned __int64") => {}
+        (&Type::PrimUnsigned(8), "unsigned char") => {}
+
+        (&Type::Ptr(&Type::PrimFloat(32)), "float*") => {}
+        (&Type::Ptr(&Type::PrimFloat(64)), "double*") => {}
+        (&Type::Ptr(&Type::PrimSigned(32)), "int*") => {}
+        (&Type::Ptr(&Type::PrimSigned(64)), "__int64*") => {}
+        (&Type::Ptr(&Type::PrimSigned(8)), "char*") => {}
+        (&Type::Ptr(&Type::PrimUnsigned(8)), "const void*") => {}
+        (&Type::Ptr(&Type::PrimUnsigned(8)), "void*") => {}
+
+        (&Type::Signed(a, b), "__m128i") |
+        (&Type::Unsigned(a, b), "__m128i") |
+        (&Type::Ptr(&Type::Signed(a, b)), "__m128i*") |
+        (&Type::Ptr(&Type::Unsigned(a, b)), "__m128i*") if a * b == 128 => {}
+
+        (&Type::Signed(a, b), "__m256i") |
+        (&Type::Unsigned(a, b), "__m256i") |
+        (&Type::Ptr(&Type::Signed(a, b)), "__m256i*") |
+        (&Type::Ptr(&Type::Unsigned(a, b)), "__m256i*") if (a as u32) * (b as u32) == 256 => {}
+
+        (&Type::Signed(a, b), "__m64") |
+        (&Type::Unsigned(a, b), "__m64") |
+        (&Type::Ptr(&Type::Signed(a, b)), "__m64*") |
+        (&Type::Ptr(&Type::Unsigned(a, b)), "__m64*") if a * b == 64 => {}
+
+        (&Type::Float(32, 4), "__m128") => {}
+        (&Type::Ptr(&Type::Float(32, 4)), "__m128*") => {}
+
+        (&Type::Float(64, 2), "__m128d") => {}
+        (&Type::Ptr(&Type::Float(64, 2)), "__m128d*") => {}
+
+        (&Type::Float(32, 8), "__m256") => {}
+        (&Type::Ptr(&Type::Float(32, 8)), "__m256*") => {}
+
+        (&Type::Float(64, 4), "__m256d") => {}
+        (&Type::Ptr(&Type::Float(64, 4)), "__m256d*") => {}
+
+        // These two intrinsics return a 16-bit element but in Intel's
+        // intrinsics they're listed as returning an `int`.
+        (&Type::PrimSigned(16), "int") if intrinsic == "_mm_extract_pi16" => {}
+        (&Type::PrimSigned(16), "int") if intrinsic == "_m_pextrw" => {}
+
+        // This intrinsic takes an `i8` to get inserted into an i8 vector, but
+        // Intel says the argument is i32...
+        (&Type::PrimSigned(8), "int") if intrinsic == "_mm_insert_epi8" => {}
+
+        // This is a macro (?) in C which seems to mutate its arguments, but that
+        // means that we're taking pointers to arguments in rust as we're not
+        // exposing it as a macro.
+        (&Type::Ptr(&Type::Float(32, 4)), "__m128") if intrinsic == "_MM_TRANSPOSE4_PS" => {}
+
+        // These intrinsics return an `int` in C but they're always either the
+        // bit 1 or 0 so we switch it to returning `bool` in rust
+        (&Type::Bool, "int")
+            if intrinsic.starts_with("_mm_comi") && intrinsic.ends_with("_sd")
+            => {}
+        (&Type::Bool, "int")
+            if intrinsic.starts_with("_mm_ucomi") && intrinsic.ends_with("_sd")
+                => {}
+
+        _ => panic!("failed to equate: `{}` and {:?} for {}", intel, t, intrinsic),
+    }
+}
```

Comment:
Nice work here. 

The two main points that worry me a bit is that:

- this equates `int` to 32-bit integer, but an `int` in C can be a 64-bit integer (and actually are in some platforms). I think assuming `int` to mean 32-bit integer is fine, but
- Intel often writes that the argument is an 8-bit integer type, even though the argument type is `int`, or the specification mentions that the value is at most N, meaning that a 2-bit, 4-bit, 8-bit, 16-bit integer types might make more sense to represent these,
- this assumes that `int` maps to a signed integer, but in many intrinsics intel uses `int` to denote a count or a bit index, which can't never be negative.

Reply:
> but an int in C can be a 64-bit integer (and actually are in some platforms). 

Not on Intel AFAIK.

Reply:
It doesn't depend on the CPU vendor, but on the platform (which must conform to the C standard). For example: 

- on `x86_64-unknown-linux-gnu` `int`s are 32-bit, and long are `64-bit`, 
- on `x86_64-pc-windows-msvc` `int`s and `long`s are 32-bit and `long long`s are 64-bit, and
- on `x86_64-cray-linux-gnu` `int`s are also 64-bit. 

All of these conform to the C standard.

Reply:
Er what I mean is this is that this doesn't matter for us. Rust doesn't have any platforms where i32 isn't `int`, so we can just cross this bridge when we get there. There's no real need to add hundreds of lines of "portability code" that aren't actually buying us anything.